hoomd/md/TwoStepNVTAlchemy.cc

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#include "TwoStepNVTAlchemy.h"
+#include "hoomd/VectorMath.h"
+#include "hoomd/md/AlchemostatTwoStep.h"
+
+namespace py = pybind11;
+
+/*! \file TwoStepNVTAlchemy.cc
+    \brief Contains code for the TwoStepNVTAlchemy class
+*/
+
+namespace hoomd
+    {
+
+namespace md
+    {
+
+/*! \param sysdef SystemDefinition this method will act on. Must not be NULL.
+    \param group The group of particles this integration method is to work on
+    \param skip_restart Skip initialization of the restart information
+*/
+TwoStepNVTAlchemy::TwoStepNVTAlchemy(std::shared_ptr<SystemDefinition> sysdef,
+                                     unsigned int alchemTimeFactor,
+                                     std::shared_ptr<Variant> T)
+    : AlchemostatTwoStep(sysdef, alchemTimeFactor), m_Q(1.0), m_alchem_KE(0.0), m_T(T)
+    {
+    m_exec_conf->msg->notice(5) << "Constructing TwoStepNVTAlchemy" << std::endl;
+
+    m_thermostat.xi = 0;
+    m_thermostat.eta = 0;
+
+#ifdef ENABLE_MPI
+    if (this->m_sysdef->isDomainDecomposed())
+        {
+        // This code is not tested or validated with MPI.
+        throw std::runtime_error("Alchemical NVT integration method does not support MPI.");
+        }
+#endif
+    }
+
+TwoStepNVTAlchemy::~TwoStepNVTAlchemy()
+    {
+    m_exec_conf->msg->notice(5) << "Destroying TwoStepNVTAlchemy" << std::endl;
+    }
+
+void TwoStepNVTAlchemy::integrateStepOne(uint64_t timestep)
+    {
+    if (timestep != m_nextAlchemTimeStep)
+        return;
+
+    m_exec_conf->msg->notice(10) << "TwoStepNVTAlchemy: 1st Alchemcial Half Step" << std::endl;
+
+    m_nextAlchemTimeStep += m_nTimeFactor;
+
+    m_alchem_KE = Scalar(0);
+
+    Scalar dUextdalpha = Scalar(0);
+
+    m_validState = false;
```

Comment:
Is this necessary as this is only set in the `integrateStep{One,Two}`.

Reply:
I don't know what the original author intended with this. IntegratorTwoStep clearly defines the two step protocol where it calls step one, computes forces, then calls step two and only calls them once per timestep. `m_validState` appears to be guarding against the impossibility that step two is called twice on the same step. 

I removed it.